# JSON Based Data Exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
****

### Import Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import json
from pandas.io.json import json_normalize

In [2]:
#Load JSON dataset in with pd.read_json()
projects_df = pd.read_json('data/world_bank_projects.json')
projects_df.head()

sector supplementprojectflg  \
0  [{'Name': 'Primary education'}, {'Name': 'Seco...                    N   
1  [{'Name': 'Public administration- Other social...                    N   
2  [{'Name': 'Rural and Inter-Urban Roads and Hig...                    Y   
3                [{'Name': 'Other social services'}]                    N   
4  [{'Name': 'General industry and trade sector'}...                    N   

  projectfinancialtype prodline  \
0                  IDA       PE   
1                OTHER       RE   
2                  IDA       PE   
3                OTHER       RE   
4                  IDA       PE   

                                             mjtheme  idacommamt  \
0                                [Human development]   130000000   
1  [Economic management, Social protection and ri...           0   
2  [Trade and integration, Public sector governan...     6060000   
3  [Social dev/gender/inclusion, Social dev/gende...           0   
4  [Trade and integration, Financial and private ...    13100000   

                                  impagency  \
0                     MINISTRY OF EDUCATION   
1                       MINISTRY OF FINANCE   
2  MINISTRY OF TRANSPORT AND COMMUNICATIONS   
3  LABOR INTENSIVE PUBLIC WORKS PROJECT PMU   
4            MINISTRY OF TRADE AND INDUSTRY   

                                        project_name mjthemecode  \
0  Ethiopia General Education Quality Improvement...        8,11   
1          TN: DTF Social Protection Reforms Support         1,6   
2  Tuvalu Aviation Investment Project - Additiona...    5,2,11,6   
3   Gov't and Civil Society Organization Partnership         7,7   
4  Second Private Sector Competitiveness and Econ...         5,4   

            closingdate  ...  \
0  2018-07-07T00:00:00Z  ...   
1                   NaN  ...   
2                   NaN  ...   
3                   NaN  ...   
4  2019-04-30T00:00:00Z  ...   

                                 majorsector_percent board_approval_month  \
0  [{'Percent': 46, 'Name': 'Education'}, {'Perce...             November   
1  [{'Percent': 70, 'Name': 'Public Administratio...             November   
2       [{'Percent': 100, 'Name': 'Transportation'}]             November   
3  [{'Percent': 100, 'Name': 'Health and other so...              October   
4  [{'Percent': 50, 'Name': 'Industry and trade'}...              October   

                                      theme_namecode  \
0      [{'code': '65', 'name': 'Education for all'}]   
1  [{'code': '24', 'name': 'Other economic manage...   
2  [{'code': '47', 'name': 'Regional integration'...   
3  [{'code': '57', 'name': 'Participation and civ...   
4  [{'code': '45', 'name': 'Export development an...   

                               countryname  \
0  Federal Democratic Republic of Ethiopia   
1                      Republic of Tunisia   
2                                   Tuvalu   
3                        Republic of Yemen   
4                       Kingdom of Lesotho   

                                                 url source  \
0  http://www.worldbank.org/projects/P129828/ethi...   IBRD   
1  http://www.worldbank.org/projects/P144674?lang=en   IBRD   
2  http://www.worldbank.org/projects/P145310?lang=en   IBRD   
3  http://www.worldbank.org/projects/P144665?lang=en   IBRD   
4  http://www.worldbank.org/projects/P144933/seco...   IBRD   

  projectstatusdisplay ibrdcommamt  \
0               Active           0   
1               Active           0   
2               Active           0   
3               Active           0   
4               Active           0   

                                     sector_namecode  \
0  [{'code': 'EP', 'name': 'Primary education'}, ...   
1  [{'code': 'BS', 'name': 'Public administration...   
2  [{'code': 'TI', 'name': 'Rural and Inter-Urban...   
3  [{'code': 'JB', 'name': 'Other social services'}]   
4  [{'code': 'YZ', 'name': 'General industry and ...   

                                    _id  
0  {'$oid': '5

In [3]:
projects_df.columns

Index(['sector', 'supplementprojectflg', 'projectfinancialtype', 'prodline',
       'mjtheme', 'idacommamt', 'impagency', 'project_name', 'mjthemecode',
       'closingdate', 'totalcommamt', 'id', 'mjsector_namecode', 'docty',
       'sector1', 'lendinginstr', 'countrycode', 'sector2', 'totalamt',
       'mjtheme_namecode', 'boardapprovaldate', 'countryshortname', 'sector4',
       'prodlinetext', 'productlinetype', 'regionname', 'status',
       'country_namecode', 'envassesmentcategorycode', 'project_abstract',
       'approvalfy', 'projectdocs', 'lendprojectcost', 'lendinginstrtype',
       'theme1', 'grantamt', 'themecode', 'borrower', 'sectorcode', 'sector3',
       'majorsector_percent', 'board_approval_month', 'theme_namecode',
       'countryname', 'url', 'source', 'projectstatusdisplay', 'ibrdcommamt',
       'sector_namecode', '_id'],
      dtype='object')

In [4]:
#Shape of dataset
projects_df.shape

(500, 50)

### 1. Find the 10 countries with the most projects

In [5]:
#Use a groupby to count # of projects for each country and sort in descending order for 10 countries with most projects
projects_df.groupby(['countryname']).size().sort_values(ascending=False).head(10)

countryname
People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Nepal                              12
People's Republic of Bangladesh    12
Kingdom of Morocco                 12
Africa                             11
Republic of Mozambique             11
dtype: int64

###  2. Find the top 10 major project themes (using column 'mjtheme_namecode')

In [6]:
#Review 'mjtheme_namecode' column of dataset
projects_df['mjtheme_namecode'].head()

0    [{'code': '8', 'name': 'Human development'}, {...
1    [{'code': '1', 'name': 'Economic management'},...
2    [{'code': '5', 'name': 'Trade and integration'...
3    [{'code': '7', 'name': 'Social dev/gender/incl...
4    [{'code': '5', 'name': 'Trade and integration'...
Name: mjtheme_namecode, dtype: object

In [7]:
#Load json and normalize the json string to flatten the dictionary for field 'mjtheme_namecode'
data = json.load((open('data/world_bank_projects.json')))
mjtheme_df=json_normalize(data,'mjtheme_namecode',['id'])
#Get shape to verify the shape with final dataframe in question 3
mjtheme_df.shape

(1499, 3)

In [8]:
#Get count of project themes and sort in descending order in order to get top 10 major project themes
mjtheme_df.name.value_counts().sort_values(ascending=False).head(10)

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

Note: There are 122 namecodes that are missing a project theme name.

In [9]:
#How many codes in each project
mjtheme_df.groupby(['id']).size().sort_values(ascending=False).head()

id
P129652    5
P127723    5
P128891    5
P128832    5
P128284    5
dtype: int64

Note: There are many projects associated with more than 1 project theme name.

In [10]:
#Exploration of a few projects associated with multiple project theme names.
print(mjtheme_df[mjtheme_df.id =='P127723'])

    code                                          name       id
375   11  Environment and natural resources management  P127723
376    5                         Trade and integration  P127723
377   11  Environment and natural resources management  P127723
378    4      Financial and private sector development  P127723
379    6         Social protection and risk management  P127723


Note: Project P127723 has been linked to 'Environment and naturaal resources management' twice

In [11]:
#Are there many projects with any duplicated associations?
mjtheme_df.groupby(['id','code']).size().sort_values(ascending=False).head()

id       code
P129663  8       5
P130873  10      5
P131945  8       5
P131194  8       5
P131919  8       5
dtype: int64

 Note: The above listed are 5 projects among many with duplicated associatons

### 3. Create a dataframe with the missing names filled in.

In [12]:
#Replace missing values with np.nan
mjtheme_df['name']=mjtheme_df['name'].replace('',np.nan)
#Subset the dataset to exclude id
mjtheme_df1=mjtheme_df[['code','name']]
#Get unique combination of code and name from dataset and exclude NaNs
mjtheme_df2=mjtheme_df1.drop_duplicates()
mjtheme_lookuptable=mjtheme_df2[mjtheme_df2['name'].notnull()]
#Get only codes from initial df and merge in names with code as key to get all values including missing
mjtheme_df1=pd.DataFrame(mjtheme_df1['code'])
mergedmjtheme_df=mjtheme_df1.merge(mjtheme_lookuptable,on='code',how='left')
#Verify that the shape matches original dataset
print(mergedmjtheme_df.shape)

(1499, 2)


In [13]:
#Get new count of project themes with filled in values
mj_with_dupes=pd.DataFrame(mergedmjtheme_df.name.value_counts().sort_values(ascending=False))
mj_with_dupes.columns=['with dupes']
mj_with_dupes=mj_with_dupes.reset_index()
mj_with_dupes.head(10)

index  with dupes
0  Environment and natural resources management         250
1                             Rural development         216
2                             Human development         210
3                      Public sector governance         199
4         Social protection and risk management         168
5      Financial and private sector development         146
6                   Social dev/gender/inclusion         130
7                         Trade and integration          77
8                             Urban development          50
9                           Economic management          38

Note: The missing project themes have been replaced with accurate project theme names.

### 3. Second approach to create dataframe with the missing names filled in.

In [14]:
#Subset dataset to exclude name and drop duplicate combinations of id and code
mjtheme_df_unique=mjtheme_df[['id','code']].drop_duplicates()
#Get only codes from de-duped df and merge in names with code as key to get all values including missing
mjtheme_df1_unique=pd.DataFrame(mjtheme_df_unique['code'])
mergedmjtheme_df_unique=mjtheme_df1_unique.merge(mjtheme_lookuptable,on='code',how='left')
#Verify that the shape matches original dataset
print(mergedmjtheme_df_unique.shape)

(1070, 2)


In [15]:
#Get a new count of project themes with filled in values 
mj_without_dupes=pd.DataFrame(mergedmjtheme_df_unique.name.value_counts().sort_values(ascending=False))
mj_without_dupes.columns=['without dupes']
mj_without_dupes=mj_without_dupes.reset_index()
mj_without_dupes.head(10)

index  without dupes
0  Environment and natural resources management            157
1                             Rural development            148
2                      Public sector governance            140
3                             Human development            128
4      Financial and private sector development            119
5         Social protection and risk management            116
6                   Social dev/gender/inclusion            114
7                         Trade and integration             61
8                             Urban development             40
9                           Economic management             33

In [16]:
#Combine both counts for compare/contrast analysis
mj_count_diff=mj_with_dupes.merge(mj_without_dupes,how='outer')
mj_count_diff['count_diff']=mj_count_diff['with dupes']-mj_count_diff['without dupes']
mj_count_diff

index  with dupes  without dupes  \
0   Environment and natural resources management         250            157   
1                              Rural development         216            148   
2                              Human development         210            128   
3                       Public sector governance         199            140   
4          Social protection and risk management         168            116   
5       Financial and private sector development         146            119   
6                    Social dev/gender/inclusion         130            114   
7                          Trade and integration          77             61   
8                              Urban development          50             40   
9                            Economic management          38             33   
10                                   Rule of law          15             14   

    count_diff  
0           93  
1           68  
2           82  
3           59  
4           52  
5           27  
6           16  
7           16  
8           10  
9            5  
10           1

Note: Environmental and natural resources management has the highest count difference after duplicates were removed from dataset.